### PR TITLE
Simplify ElasticsearchException rendering as a XContent

### PR DESCRIPTION
**Pull request by tlrx**
This pull request tries to simplify the way `ElasticsearchException` are rendered to xcontent. It adds some documentation and renames and merges some methods. Current behavior is preserved, the goal is to be more readable and centralize everything in the `ElasticsearchException` class.

### CODE REVIEW

File: core/src/main/java/org/elasticsearch/ElasticsearchException.java [hunk 1]
```diff
@@ -67,12 +69,13 @@
...
     public static final String REST_EXCEPTION_SKIP_STACK_TRACE = "rest.exception.stacktrace.skip";
     public static final boolean REST_EXCEPTION_SKIP_STACK_TRACE_DEFAULT = true;
-    public static final boolean REST_EXCEPTION_SKIP_CAUSE_DEFAULT = false;
-    private static final String INDEX_HEADER_KEY = "es.index";
-    private static final String INDEX_HEADER_KEY_UUID = "es.index_uuid";
-    private static final String SHARD_HEADER_KEY = "es.shard";
-    private static final String RESOURCE_HEADER_TYPE_KEY = "es.resource.type";
-    private static final String RESOURCE_HEADER_ID_KEY = "es.resource.id";
+    private static final boolean REST_EXCEPTION_SKIP_CAUSE_DEFAULT = false;
+    private static final String ES_HEADER_PREFIX = "es.";
```
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/ElasticsearchException.java]: I don't like this. the reason is that when you grep for `es.index` you find nothing. I think we should never do this.

File: core/src/main/java/org/elasticsearch/ElasticsearchException.java [hunk 2]
```diff
@@ -328,25 +319,94 @@ private void xContentHeader(XContentBuilder builder, String key, List<String> va
...
+            builder.endArray();
         }
+        toXContent(builder, params, e);
+        builder.endObject();
+    }
+
+    /**
+     * Same as {@link ElasticsearchException#toXContentError(XContentBuilder, Params, Exception, boolean)} with all details.
+     */
+    public static void toXContentError(XContentBuilder builder, Params params, @Nullable Exception e) throws IOException {
```
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/ElasticsearchException.java]: this name is not good. there is no *error* involved. I think you should maybe name it `generateFailureXContent` or somethign like this

**Review by s1monw — CHANGES_REQUESTED** [review]
left some comments

File: core/src/main/java/org/elasticsearch/ElasticsearchException.java (hunk 1 above)
Comment (tlrx) [inline core/src/main/java/org/elasticsearch/ElasticsearchException.java]: Good point

File: core/src/main/java/org/elasticsearch/ElasticsearchException.java (hunk 2 above)
Comment (tlrx) [inline core/src/main/java/org/elasticsearch/ElasticsearchException.java]: Ok

**Comment by tlrx** [comment]
@s1monw Thanks for the feedback, I updated the PR.

File: core/src/main/java/org/elasticsearch/ElasticsearchException.java [hunk 3]
```diff
@@ -260,60 +262,48 @@ public XContentBuilder toXContent(XContentBuilder builder, Params params) throws
         } else {
             builder.field(TYPE, getExceptionName());
             builder.field(REASON, getMessage());
-            for (String key : headers.keySet()) {
-                if (key.startsWith("es.")) {
-                    List<String> values = headers.get(key);
-                    xContentHeader(builder, key.substring("es.".length()), values);
+
+            Set<String> customHeaders = null;
+            for (String header : headers.keySet()) {
```
Comment (javanna) [inline core/src/main/java/org/elasticsearch/ElasticsearchException.java]: would you mind reverting the variable name change, at least temporarily? it confuses the review

Comment (tlrx) [inline core/src/main/java/org/elasticsearch/ElasticsearchException.java]: Done, sorry for the annoyance.

Comment (javanna) [inline core/src/main/java/org/elasticsearch/ElasticsearchException.java]: I didn't mean to revert the changes made after SImon's review, I meant to revert the variable rename from key to header.

File: core/src/main/java/org/elasticsearch/ElasticsearchException.java [hunk 4]
```diff
@@ -328,25 +318,95 @@ private void xContentHeader(XContentBuilder builder, String key, List<String> va
...
+        ElasticsearchException exception;
+        if (t instanceof ElasticsearchException) {
+            exception = (ElasticsearchException) t;
         } else {
-            builder.field(TYPE, getExceptionName(ex));
-            builder.field(REASON, ex.getMessage());
-            if (ex.getCause() != null) {
-                builder.field(CAUSED_BY);
+            final Throwable throwable = t;
+            exception = new ElasticsearchException(throwable.getMessage(), throwable.getCause()) {
```
Comment (javanna) [inline core/src/main/java/org/elasticsearch/ElasticsearchException.java]: with this change we now obey to the rest.exception.cause.skip also for non elasticsearch exception while we previously weren't, is that correct?

Comment (javanna) [inline core/src/main/java/org/elasticsearch/ElasticsearchException.java]: maybe instead of wrapping the exception we should have a static method that takes as an argument all the little pieces that get printed out, then that method could be called here and in toXContent?

File: core/src/main/java/org/elasticsearch/ElasticsearchException.java [hunk 5]
```diff
@@ -328,25 +318,95 @@ private void xContentHeader(XContentBuilder builder, String key, List<String> va
...
+     */
+    protected boolean skipCauseInXContent(Params params) {
+        return params.paramAsBoolean(REST_EXCEPTION_SKIP_CAUSE, REST_EXCEPTION_SKIP_CAUSE_DEFAULT);
+    }
+
+    /**
+     * Static toXContent helper method that also renders non {@link org.elasticsearch.ElasticsearchException} instances as XContent,
+     * delegating the rendering to {@link ElasticsearchException#toXContent(XContentBuilder, Params)}.
+     */
+    public static void toXContent(XContentBuilder builder, Params params, Throwable t) throws IOException {
```
Comment (javanna) [inline core/src/main/java/org/elasticsearch/ElasticsearchException.java]: I would try and rename this one as well, I find it annoying that is has the same name as the ordinary toXContent.

File: core/src/main/java/org/elasticsearch/ElasticsearchException.java [hunk 6]
```diff
@@ -328,25 +318,95 @@ private void xContentHeader(XContentBuilder builder, String key, List<String> va
...
+            builder.endArray();
         }
+        toXContent(builder, params, e);
+        builder.endObject();
+    }
+
+    /**
+     * Same as {@link ElasticsearchException#generateFailureXContent(XContentBuilder, Params, Exception, boolean)} with all details.
+     */
+    public static void generateFailureXContent(XContentBuilder builder, Params params, @Nullable Exception e) throws IOException {
```
Comment (javanna) [inline core/src/main/java/org/elasticsearch/ElasticsearchException.java]: maybe we can live without this one?

File: core/src/main/java/org/elasticsearch/ElasticsearchException.java [hunk 7]
```diff
@@ -328,25 +318,95 @@ private void xContentHeader(XContentBuilder builder, String key, List<String> va
...
+        exception.toXContent(builder, params);
+    }
+
+    /**
+     * Render any exception as a xcontent, encapsulated within a field or object named "error". The level of details that are rendered
+     * depends on the value of the "detailed" parameter: when it equals to false, only a simple message based on the type and message of the
+     * exception is rendered. When it equals to true, all detail are provided including guesses root causes, cause and potentially stack
+     * trace.
+     */
+    public static void generateFailureXContent(XContentBuilder builder, Params params, @Nullable Exception e, boolean detailed)
```
Comment (javanna) [inline core/src/main/java/org/elasticsearch/ElasticsearchException.java]: nice! so this one is the previous renderException, but it also contains the bits coming from BytesRestResponse#convert, right? But that is fine because the users of renderException will always use the detailed part and provide a non null exception?

File: core/src/test/java/org/elasticsearch/ElasticsearchExceptionTests.java [hunk 8]
```diff
@@ -67,7 +67,7 @@ public void testToXContent() throws IOException {
         // in the JSON. Since the stack can be large, it only checks the beginning of the JSON.
         assertExceptionAsJson(e, true, startsWith("{\"type\":\"exception\",\"reason\":\"foo\"," +
                 "\"caused_by\":{\"type\":\"illegal_state_exception\",\"reason\":\"bar\"," +
-                "\"stack_trace\":\"java.lang.IllegalStateException: bar"));
+                "\"stack_trace\":\"[bar]\\n\\tat org.elasticsearch."));
```
Comment (javanna) [inline core/src/test/java/org/elasticsearch/ElasticsearchExceptionTests.java]: what happened here?

File: core/src/main/java/org/elasticsearch/action/search/SearchPhaseExecutionException.java [hunk 9]
```diff
@@ -144,15 +144,15 @@ protected void innerToXContent(XContentBuilder builder, Params params) throws IO
...
-    protected void causeToXContent(XContentBuilder builder, Params params) throws IOException {
-        if (super.getCause() != null) {
-            // if the cause is null we inject a guessed root cause that will then be rendered twice so wi disable it manually
-            super.causeToXContent(builder, params);
+    protected boolean skipCauseInXContent(Params params) {
+        if (super.skipCauseInXContent(params)) {
+            return true;
         }
+        // if the cause is null we inject a guessed root cause that will then be rendered twice so we disable it manually
+        return super.getCause() == null;
```
Comment (javanna) [inline core/src/main/java/org/elasticsearch/action/search/SearchPhaseExecutionException.java]: I don't follow the logic here and I wish we wouldn't need to have this skipCauseInXContent protected method. How does the behaviour here differ from the default one?

File: core/src/main/java/org/elasticsearch/ElasticsearchException.java [hunk 10]
```diff
@@ -328,25 +318,95 @@ private void xContentHeader(XContentBuilder builder, String key, List<String> va
...
 
     /**
-     * Static toXContent helper method that also renders non {@link org.elasticsearch.ElasticsearchException} instances as XContent.
+     * Renders additional per exception information into the XContent
      */
-    public static void toXContent(XContentBuilder builder, Params params, Throwable ex) throws IOException {
-        ex = ExceptionsHelper.unwrapCause(ex);
-        if (ex instanceof ElasticsearchException) {
-            ((ElasticsearchException) ex).toXContent(builder, params);
+    protected void innerToXContent(XContentBuilder builder, Params params) throws IOException {
```
Comment (javanna) [inline core/src/main/java/org/elasticsearch/ElasticsearchException.java]: shall we rename this too to highlight that it's for additional info?

File: core/src/main/java/org/elasticsearch/ElasticsearchException.java [hunk 11]
```diff
@@ -260,60 +262,48 @@ public XContentBuilder toXContent(XContentBuilder builder, Params params) throws
...
-            xContentHeader(builder, key, values);
-        }
-        if (hasHeader) {
-            builder.endObject();
         }
+        return builder;
     }
 
-    private void xContentHeader(XContentBuilder builder, String key, List<String> values) throws IOException {
+    private void headerToXContent(XContentBuilder builder, String key, List<String> values) throws IOException {
```
Comment (javanna) [inline core/src/main/java/org/elasticsearch/ElasticsearchException.java]: static?

File: core/src/main/java/org/elasticsearch/ElasticsearchException.java [hunk 12]
```diff
@@ -328,25 +318,95 @@ private void xContentHeader(XContentBuilder builder, String key, List<String> va
...
-                builder.field(CAUSED_BY);
+            final Throwable throwable = t;
+            exception = new ElasticsearchException(throwable.getMessage(), throwable.getCause()) {
+                @Override
+                protected String getExceptionName() {
+                    return getExceptionName(throwable);
+                }
+            };
+        }
+        assert exception != null : "elasticsearch exception can't be null";
```
Comment (javanna) [inline core/src/main/java/org/elasticsearch/ElasticsearchException.java]: this assert doesn't make sense here? it can never be null?

**Review by javanna — CHANGES_REQUESTED** [review]
I left some comments. I like the idea.

File: core/src/main/java/org/elasticsearch/ElasticsearchException.java [hunk 13]
```diff
@@ -260,60 +262,48 @@ public XContentBuilder toXContent(XContentBuilder builder, Params params) throws
...
-        }
-    }
+            innerToXContent(builder, params);
 
-    protected final void renderHeader(XContentBuilder builder, Params params) throws IOException {
-        boolean hasHeader = false;
-        for (String key : headers.keySet()) {
-            if (key.startsWith("es.")) {
-                continue;
+            boolean skipCause = skipCauseInXContent(params);
```
Comment (tlrx) [inline core/src/main/java/org/elasticsearch/ElasticsearchException.java]: No, let's remove it

File: core/src/main/java/org/elasticsearch/ElasticsearchException.java (hunk 10 above)
Comment (tlrx) [inline core/src/main/java/org/elasticsearch/ElasticsearchException.java]: Something like `metadataToXContent` (similar to `headerToXContent`)?

File: core/src/main/java/org/elasticsearch/ElasticsearchException.java (hunk 5 above)
Comment (tlrx) [inline core/src/main/java/org/elasticsearch/ElasticsearchException.java]: I have no idea how to rename this method, any suggestion? `generateThrowableXContent` ?

File: core/src/main/java/org/elasticsearch/ElasticsearchException.java (hunk 10 above)
Comment (javanna) [inline core/src/main/java/org/elasticsearch/ElasticsearchException.java]: ++

File: core/src/main/java/org/elasticsearch/ElasticsearchException.java (hunk 5 above)
Comment (javanna) [inline core/src/main/java/org/elasticsearch/ElasticsearchException.java]: that sounds good yes

File: core/src/test/java/org/elasticsearch/ElasticsearchExceptionTests.java (hunk 8 above)
Comment (tlrx) [inline core/src/test/java/org/elasticsearch/ElasticsearchExceptionTests.java]: The cause was wrapped within the (now removed) anonymous ElasticsearchException so it was rendered using ElasticsearchException.toString() which was wrong. I should not have modified this.

**Comment by tlrx** [comment]
@javanna I updated the PR according to your comments. Can you have another look please?

**Comment by tlrx** [comment]
@elasticmachine retest this please

**Comment by tlrx** [comment]
@elasticmachine retest this please

File: core/src/main/java/org/elasticsearch/ElasticsearchException.java (hunk 6 above)
Comment (javanna) [inline core/src/main/java/org/elasticsearch/ElasticsearchException.java]: shall we remove it?

File: core/src/main/java/org/elasticsearch/action/search/SearchPhaseExecutionException.java [hunk 14]
```diff
@@ -131,7 +131,7 @@ private static String buildMessage(String phaseName, String msg, ShardSearchFail
     }
 
     @Override
```
Comment (javanna) [inline core/src/main/java/org/elasticsearch/action/search/SearchPhaseExecutionException.java]: can you add `@Override` to the `getCause` method in this class please?

File: core/src/main/java/org/elasticsearch/action/search/SearchPhaseExecutionException.java [hunk 15]
```diff
@@ -144,15 +144,20 @@ protected void innerToXContent(XContentBuilder builder, Params params) throws IO
...
-            // if the cause is null we inject a guessed root cause that will then be rendered twice so wi disable it manually
-            super.causeToXContent(builder, params);
+    public XContentBuilder toXContent(XContentBuilder builder, Params params) throws IOException {
+        Throwable ex = ExceptionsHelper.unwrapCause(this);
+        if (ex != this) {
+            generateThrowableXContent(builder, params, this);
+        } else {
+            // We don't have a cause when all shards failed, but we do have shards failures so we can "guess" a cause.
+            // Here, we use super.getCause() because we don't want the guessed exception to be rendered twice (one in
+            // the "cause" field, one in "failed_shards")
```
Comment (javanna) [inline core/src/main/java/org/elasticsearch/action/search/SearchPhaseExecutionException.java]: thanks for changing the explanation! makes more sense now. Can you also point it to the getCause method where we guess the cause?

File: core/src/test/java/org/elasticsearch/ElasticsearchExceptionTests.java [hunk 16]
```diff
@@ -67,7 +67,7 @@ public void testToXContent() throws IOException {
         // in the JSON. Since the stack can be large, it only checks the beginning of the JSON.
         assertExceptionAsJson(e, true, startsWith("{\"type\":\"exception\",\"reason\":\"foo\"," +
                 "\"caused_by\":{\"type\":\"illegal_state_exception\",\"reason\":\"bar\"," +
-                "\"stack_trace\":\"java.lang.IllegalStateException: bar"));
+                "\"stack_trace\":\"java.lang.IllegalStateException: bar\\n\\tat org.elasticsearch."));
```
Comment (javanna) [inline core/src/test/java/org/elasticsearch/ElasticsearchExceptionTests.java]: did we change the output here or did we just improve the test?

File: core/src/main/java/org/elasticsearch/rest/BytesRestResponse.java [hunk 17]
```diff
@@ -116,31 +119,23 @@ public RestStatus status() {
...
-            ElasticsearchException.renderException(builder, params, e);
-        } else {
-            builder.field("error", simpleMessage(e));
         }
+
+        XContentBuilder builder = channel.newErrorBuilder().startObject();
+        ElasticsearchException.generateFailureXContent(builder, params, e, channel.detailedErrorsEnabled());
         builder.field("status", status.getStatus());
         builder.endObject();
         return builder;
```
Comment (javanna) [inline core/src/main/java/org/elasticsearch/rest/BytesRestResponse.java]: the simpleMessage method is still in this class but unused, I think it should be removed

File: core/src/main/java/org/elasticsearch/ElasticsearchException.java [hunk 18]
```diff
@@ -328,25 +327,75 @@ private void xContentHeader(XContentBuilder builder, String key, List<String> va
...
      */
-    public static void toXContent(XContentBuilder builder, Params params, Throwable ex) throws IOException {
-        ex = ExceptionsHelper.unwrapCause(ex);
-        if (ex instanceof ElasticsearchException) {
-            ((ElasticsearchException) ex).toXContent(builder, params);
+    protected void metadataToXContent(XContentBuilder builder, Params params) throws IOException {
+    }
+
+    /**
+     * Static toXContent helper method that also renders non {@link org.elasticsearch.ElasticsearchException} instances as XContent,
```
Comment (javanna) [inline core/src/main/java/org/elasticsearch/ElasticsearchException.java]: can you clarify in the javadocs that it supports Throwable besides what you already wrote?

File: core/src/main/java/org/elasticsearch/ElasticsearchException.java [hunk 19]
```diff
@@ -328,25 +327,75 @@ private void xContentHeader(XContentBuilder builder, String key, List<String> va
...
-            builder.field(REASON, ex.getMessage());
-            if (ex.getCause() != null) {
-                builder.field(CAUSED_BY);
+            innerToXContent(builder, params, t, getExceptionName(t), t.getMessage(), emptyMap(), t.getCause());
+        }
+    }
+
+    /**
+     * Render any exception as a xcontent, encapsulated within a field or object named "error". The level of details that are rendered
+     * depends on the value of the "detailed" parameter: when it equals to false, only a simple message based on the type and message of the
```
Comment (javanna) [inline core/src/main/java/org/elasticsearch/ElasticsearchException.java]: when it's false, and later when it's true?

**Review by javanna — CHANGES_REQUESTED** [review]
I like it a lot, much cleaner than before. I am still confused on which method should be used to render exceptions, maybe we could clarify that in the javadocs to distinguish between generateFailureXContent, generateThrowableXContent (just an idea, not sure). I left a few comments but it's very close.

File: core/src/main/java/org/elasticsearch/rest/BytesRestResponse.java [hunk 20]
```diff
@@ -116,31 +119,23 @@ public RestStatus status() {
...
             }
-            ElasticsearchException.renderException(builder, params, e);
-        } else {
-            builder.field("error", simpleMessage(e));
         }
+
+        XContentBuilder builder = channel.newErrorBuilder().startObject();
+        ElasticsearchException.generateFailureXContent(builder, params, e, channel.detailedErrorsEnabled());
         builder.field("status", status.getStatus());
         builder.endObject();
```
Comment (javanna) [inline core/src/main/java/org/elasticsearch/rest/BytesRestResponse.java]: it's unfortunate that we still have the status here, it means that when writing parsing code `failureFromXContent` and `throwableFromXContent` won't be enough, we will still need a parse method as part of this class.

I wonder if we can remove the `public BytesRestResponse(RestChannel channel, RestStatus status, Exception e) throws IOException {` constructor, then the status always comes from the exception and this bit can be moved out too?

File: core/src/test/java/org/elasticsearch/ElasticsearchExceptionTests.java (hunk 16 above)
Comment (s1monw) [inline core/src/test/java/org/elasticsearch/ElasticsearchExceptionTests.java]: not sure if that will work cross platform either

File: core/src/main/java/org/elasticsearch/rest/BytesRestResponse.java (hunk 17 above)
Comment (tlrx) [inline core/src/main/java/org/elasticsearch/rest/BytesRestResponse.java]: Leftover, thanks!

File: core/src/test/java/org/elasticsearch/ElasticsearchExceptionTests.java (hunk 16 above)
Comment (tlrx) [inline core/src/test/java/org/elasticsearch/ElasticsearchExceptionTests.java]: > did we change the output here or did we just improve the test?

A little improvement to the test.

> not sure if that will work cross platform either

Right, I changed that, it should be ok on Windows now.

File: core/src/main/java/org/elasticsearch/rest/BytesRestResponse.java (hunk 20 above)
Comment (tlrx) [inline core/src/main/java/org/elasticsearch/rest/BytesRestResponse.java]: I agree it's unfortunate, specially in the context of `BytesRestResponse` or `MultiSearchResponse`. 

But on the other hand a `status` field is not always associated to the exception rendered using `ElasticsearchException.generateFailureXContent()` like in `MultiSearchTemplateResponse`, so would it make sense to add rendering logic of the status in `generateFailureXContent` and parsing logic in `failureFromXContent`?

Same remark applies for `generateThrowableXContent`, status is sometime rendered as part of the enclosing object (see DefaultShardOperationFailedException or BulkItemResponse) so custom parsing logic will be needed anyway.

But I don't have a string feeling here. I feel like we shouldn't change the current API to make it uniform as it is not clear how the `status` value associated to `ElasticsearchException` is derived (looks like `ExceptionsHelper.status(Throwable)` is used most of the time but it looks like the status value is sometimes excplicitly set to something else).

Comment (javanna) [inline core/src/main/java/org/elasticsearch/rest/BytesRestResponse.java]: I had missed that there were more constructor, we cannot work around this at the moment, thanks for looking this.

**Review by javanna — APPROVED** [review]
LGTM

**Comment by tlrx** [comment]
@s1monw Do you want to have another look?

File: core/src/main/java/org/elasticsearch/ElasticsearchException.java [hunk 21]
```diff
@@ -329,25 +328,73 @@ private void xContentHeader(XContentBuilder builder, String key, List<String> va
...
+    }
+
+    /**
+     * Static toXContent helper method that renders {@link org.elasticsearch.ElasticsearchException} or {@link Throwable} instances
+     * as XContent, delegating the rendering to {@link #toXContent(XContentBuilder, Params)}
+     * or {@link #innerToXContent(XContentBuilder, Params, Throwable, String, String, Map, Throwable)}.
+     *
+     * This method is usually used when the {@link Throwable} is rendered as a part of another XContent object.
+     */
+    public static void generateThrowableXContent(XContentBuilder builder, Params params, Throwable t) throws IOException {
```
Comment (s1monw) [inline core/src/main/java/org/elasticsearch/ElasticsearchException.java]: isn't the name implicit by the parameter list? What's wrong with `toXContent`?

**Review by s1monw — APPROVED** [review]
left a question LGTM in general

File: core/src/main/java/org/elasticsearch/ElasticsearchException.java (hunk 21 above)
Comment (javanna) [inline core/src/main/java/org/elasticsearch/ElasticsearchException.java]: I find it confusing with the `toXContent` coming from the `ToXContent` interface. Arguments help but I think naming is better now.

Comment (tlrx) [inline core/src/main/java/org/elasticsearch/ElasticsearchException.java]: I prefer the new name too.

**Comment by tlrx** [comment]
@s1monw @javanna Thanks! This can go to 5.3 as well.

**Comment by javanna** [comment]
> This can go to 5.3 as well.

++ thank you
